### PR TITLE
Check disk space before parsing, display a parse error on the frontend

### PR DESCRIPTION
**Pull request by caleblf**
Addresses #98

Changes:
- If there is less than a hardcoded minimum amount of space left on disk (both `/` and `/data/db` are checked, since they can be on different partitions), parsing is not attempted and the file is marked as an error so it can be rerun.
- If there are any parse errors, one message is displayed on the frontend in a collapsible details element.
  - This applies to a few other types of error as well, such as dask issues, and parser failures when marked as `mustSucceed`

Screenshots of the error details (the error progress bar flashes, so it's not visible in two of these images):
- Screenshot 2023-08-22 at 11 52 19
- Tested by modifying the disk space minimum; in real usage this number will be in the kilobytes rather than the gigabytes, so displaying it in bytes is more reasonable:
  Screenshot 2023-08-22 at 11 57 39
- Screenshot 2023-08-22 at 11 52 11

### CODE REVIEW

File: faw/pdf-observatory/ui/src/components/AnalysisSetConfig.vue [hunk 1]
```diff
@@ -106,7 +106,16 @@
       div
         span
           v-icon mdi-help-rhombus
-        span(style="font-size:0.85em") Pipelines attached to analysis sets are a two-stage process: 1) The pipeline is initiated. All files *currently* matching the analysis set definition, ignoring max documents, will be available to the pipeline. Thus, best to wait until all non-pipeline parsers have finished. 2) When the pipeline finishes all of its tasks, and if the pipeline specifies any parsers, all analysis sets which have specified that they will use this pipeline's parsers will have their files processed with the parsers learned from this analysis set.
+        span(style="font-size:0.85em")
```
Comment (caleblf) [inline faw/pdf-observatory/ui/src/components/AnalysisSetConfig.vue:109]: Unrelated style change (wrapping a long line) that I happened to spot while searching the UI code.

File: faw/pdf-observatory/faw_analysis_set_parse.py [hunk 2]
```diff
@@ -35,6 +34,13 @@ def _import_from_path(mod_name, pth):
 pdf_etl_parse = _import_from_path('pdf_etl_parse', '../pdf-etl-parse/main.py')
 
 COL_NAME = 'as_parse'
+"""Name of parse queue collection."""
+MIN_REQUIRED_PARSE_SPACE_BYTES = 2 * 1024 * 1024  # 2 MB
```
Comment (caleblf) [inline faw/pdf-observatory/faw_analysis_set_parse.py:38]: For now I set a 2MB minimum; this seems like overkill for any sane parsing case I can think of, but maybe we want to go higher since this could be cutting it close for comfortable ssh access.

File: faw/pdf-observatory/faw_analysis_set_parse.py [hunk 3]
```diff
@@ -264,6 +276,11 @@ def _dask_as_parse_file(app_config, api_info, doc):
     doc_invname = f'/home/pdf-files/{doc_id}'
     parsers_done = {}  # {name: [tool, parser]}
 
+    # TODO could get a more precise estimate for required space by
+    # checking size of prior parse output, considering the number of
+    # parsers being rerun, etc
+    ensure_sufficient_disk_space(MIN_REQUIRED_PARSE_SPACE_BYTES, doc_id)
```
Comment (caleblf) [inline faw/pdf-observatory/faw_analysis_set_parse.py:282]: Some notes in the comment here on potential improvements.

File: faw/pdf-observatory/faw_analysis_set_parse.py (hunk 2 above)
Comment (wwoods) [reply faw/pdf-observatory/faw_analysis_set_parse.py:38]: We should probably start with 2GB

File: faw/pdf-observatory/faw_analysis_set_parse.py [hunk 4]
```diff
@@ -695,3 +712,20 @@ def _get_cfg(k, app_config):
...
+    Raises:
+        An appropriate error if there isn't enough disk space to parse a file
+    """
+    for path in ['/', '/data/db']:  # These might be on different partitions
+        bytes_available = psutil.disk_usage(path).free
+        if bytes_available < bytes_needed:
+            # Raise a RuntimeError instead of an OSError(ENOSPC) since those
+            # technically require OS-sourced messages.
+            raise RuntimeError(
+                f'Only {bytes_available}B left on {path}; file {filename} skipped'
```
Comment (wwoods) [inline faw/pdf-observatory/faw_analysis_set_parse.py:730]: Convert error message to MB

**Review by wwoods — APPROVED** [review]
LGTM
